### DISCUSSION TITLE
Custom values for options in user profile validation

### DISCUSSION THREAD

**Discussion by LogaMithran** [keycloak/keycloak#29957] [Q&A]
I'm creating a user attribute (attribute_1 it will be a dropdown having some values) in a realm, I was able to create the attribute and add validations as options to it. 

1. Once the validators is created, I'm not able to edit that which is drawback. 
2. I need the values to be fetched dynamically from an external Api/Db (Is this possible with any SPI?)

Version - 24.0.4
![Screenshot 2024-05-29 11:33:34](https://github.com/keycloak/keycloak/assets/85341468/9ce3499b-6e45-4f13-8c6d-64a6a6a28148)

**Comment by LogaMithran** [comment]
Could someone help me in this, I have tried to manipulate the DOM like in the screenshot. But it's not storing in the DB as expected.
I hope there's a better way to do this.
![Screenshot 2024-05-31 00:03:58](https://github.com/keycloak/keycloak/assets/85341468/aa4ff151-4228-4884-9942-fa40ec12e525)

**Comment by danielFesenmeyer** [comment]
It's not easily possible. We have implemented something similar for Keycloak 23.x, and it was lots of effort. Wouldn't do it again, unless it was absolutely necessary.

It's because the options validation is tightly coupled to the validator config in several places. Would be easier when the options would be an overwritable method, with reading from validator config as default.

**Reply by LogaMithran** [reply]
How did you achieve it?, By creating custom provider or some other way, In my case it's necessary.

**Reply by danielFesenmeyer** [reply]
What we did in a nutshell:

1. Created our own Validator, e.g. `CustomOptionsValidator extends AbstractStringValidator`. For this validator, we also implemented the interface `ConfiguredProvider` for some further config options (in our case it was defaultOptions, in case the API/DB does not provide anything). In method `getId` we provided an ID like `customOptionsValidator`.
2. In the validator, we provided a method like `getCustomValidationOptions`, returning the options from API/DB.
3. In the validator's `doValidate`, we use `getCustomValidationOptions` to check whether a given value is valid (instead of reading from config, as done in the corresponding method of Keycloak's `OptionValidator`).
4. We extended FreemarkerLoginFormsProvider, and overwrote method `processTemplate`, checking whether `attributes.get("profile")` is an instance of `AbstractUserProfileBean`. If yes, we iterate over all attributes (`profileBean.getAttributes()`), and get the validator of ID `customOptionsValidator` (`attribute.getValidators("customOptionsValidator")`). If the attribute contains such a validator, we create an instance of CustomOptionsValidator, calling Validators.validator(session, "customOptionsValidator") and call method `getCustomValidationOptions`, in order to retrieve the supported options. We then patch the attribute's field `AttributeMetadata metadata`, by means of reflection, replacing `customOptionsValidator` with (Keycloak builtin) `options` validator, with a dynamically created `options` config property based on `getCustomValidationOptions`.

Note: This way, the options are displayed nicely in the login theme, but not in the Admin UI. At least, the validation works fine both in login theme and Admin UI. For the Admin UI, you would probably have to patch some React code.

**Reply by LogaMithran** [reply]
Thanks for detailed explanation Daniel. User profile feature comes by default in 24.x. I will try that point 4 in with this version.
But one thing, while fetching metadata in the user creation, tweaking my CustomOptionsValidator will help?.

**Reply by danielFesenmeyer** [reply]
Without point 4, validation will work - but in the login theme, no select box will be displayed, just a text field. As a workaround, you could tweak the validation message to show all the valid options.

In case you are just using the Admin API for creating users, point 4 is not relevant. It's just a usability enhancement (selectbox instead of textfield) for the login theme "user self-service" functionality (user registration, profile update, ...).

**Reply by LogaMithran** [reply]
Yeah, I got your point on validating the input but how can I show the values in the drop down. Since all the metadata are being fetched in this endpoint `users/profile/metadata`. 
I need to overide the values in the options. once the user creation page is loaded metadata are being fetched from here `public UserProfileMetadata getMetadata() {}`. I cannot tweak my validator in here right?.
But I could get the additional values if I use the PUT in this endpoint `users/profile` separately.

**Reply by danielFesenmeyer** [reply]
OK, it seems that something has changed from Keycloak 23.x to Keycloak 24.x.

For Keycloak 23.x, no endpoint `users/profile/metadata` was involved for showing the dropdown, the code was just in FreeMarkerLoginFormsProvider and in FreeMarker templates.
We have not yet migrated to Keycloak 24.x, hence I don't know further details :confused:

**Reply by LogaMithran** [reply]
Okay Daniel, I'll let you know if I find anything on this. Thanks for your time :)
